stats/condition_interceptor.go

```diff
@@ -0,0 +1,185 @@
+			return
+		}
+
+		if !conditionIsAbated(condition) {
+			patient := result.(*models.Patient)
+			err = s.PgDataAccess.AddConditionStat(patient, condition)
+		}
+
+		if err != nil {
+			log.Printf("ConditionStatsCreateInterceptor: %s\n", err.Error())
+		}
+	}
+}
+
+// OnError is unused
+func (s *ConditionStatsCreateInterceptor) OnError(err error, resource interface{}) {}
+
+// ConditionStatsUpdateInterceptor intercepts any updates to condition resources in the database
+// and updates the Synthetic Mass condition statistics based on the condition's patient's address.
+type ConditionStatsUpdateInterceptor struct {
+	PgDataAccess    StatsDataAccess
+	MongoDataAccess server.DataAccessLayer
+	// The state of the condition before the database update, for comparison after the database update
+	conditionsBefore map[string]*models.Condition
+	// Tracks if the interceptor failed to cache the condition model before the update
+	cacheError error
+}
+
+// NewConditionStatsUpdateInterceptor returns an initialized instance of ConditionStatsUpdateInterceptor
+func NewConditionStatsUpdateInterceptor(pgDataAccess StatsDataAccess, mongoDataAccess server.DataAccessLayer) *ConditionStatsUpdateInterceptor {
+	interceptor := &ConditionStatsUpdateInterceptor{PgDataAccess: pgDataAccess, MongoDataAccess: mongoDataAccess}
+	interceptor.conditionsBefore = make(map[string]*models.Condition)
+	return interceptor
+}
+
+// Before caches a condition resource before it's updated, for comparison after the update.
+func (s *ConditionStatsUpdateInterceptor) Before(resource interface{}) {
+	condition, ok := resource.(*models.Condition)
+
+	if ok {
+		s.conditionsBefore[condition.Id] = condition
+	} else {
+		errmsg := "ConditionStatsUpdateInterceptor:Before: Failed to cache condition before update"
+		s.cacheError = errors.New(errmsg)
+		log.Println(errmsg)
+	}
+}
+
+// After compares the updated condition resource to the cached condition resource (from before the update), then
+// updates population statistics based on that resource's patient's address.
+func (s *ConditionStatsUpdateInterceptor) After(resource interface{}) {
+	condition, ok := resource.(*models.Condition)
+
+	if ok && s.cacheError == nil {
+		// see if the condition is now abated, so we can stop tracking stats for it
+		if !conditionIsAbated(s.conditionsBefore[condition.Id]) && conditionIsAbated(condition) {
+
+			if condition.Subject == nil {
+				log.Printf("ConditionStatsUpdateInterceptor: Condition %s has no subject\n", condition.Id)
+				return
+			}
+
+			result, err := s.MongoDataAccess.Get(condition.Subject.ReferencedID, "Patient")
+
+			if err != nil {
+				log.Printf("ConditionStatsUpdateInterceptor: Failed to get patient for condition %s\n", condition.Id)
+				return
+			}
+
+			patient := result.(*models.Patient)
+			err = s.PgDataAccess.RemoveConditionStat(patient, condition)
+			delete(s.conditionsBefore, condition.Id)
```

Comment:
The problem with deleting this here is that you have introduced a potential memory leak (by uncleaned map items).  In this case where the condition did not abate, you never get into this `if` scope, so the old condition never gets deleted from the map.

That's why my suggested code above deletes it at a higher scope.